src/Foundation/NSUrlSessionHandler.cs

```diff
@@ -253,21 +253,24 @@ protected override async Task<HttpResponseMessage> SendAsync (HttpRequestMessage
 
 			cancellationToken.Register (() => {
 				RemoveInflightData (dataTask);
+				dataTask?.Cancel ();
 				tcs.TrySetCanceled ();
 			});
-
-			lock (inflightRequestsLock)
-				inflightRequests.Add (dataTask, new InflightData {
-					RequestUrl = request.RequestUri.AbsoluteUri,
-					CompletionSource = tcs,
-					CancellationToken = cancellationToken,
-					Stream = new NSUrlSessionDataTaskStream (),
-					Request = request
-				});
-
-			if (dataTask.State == NSUrlSessionTaskState.Suspended)
-				dataTask.Resume ();
-
+			
+			if (!cancellationToken.IsCancellationRequested) {
```

Comment:
That seems to always be hit (e.g. adding an `else` and a breakpoint or some `Console.WriteLine` is never hit).

Also I can reproduce the _hang_ locally, with the supplied test case, even with the patch applied. Now it works, some times, but it's hard to say if the different timings makes it more or less likely (or if it's purely random).

Considering the `Cancel` is a no-op I do not think this change is fixing the issue.

Reply:
What is always hit? I use the reproduce steps provided in the referred bug and with this it does not happens. Be sure to use a small timeout btw

Reply:
```
if (!cancellationToken.IsCancellationRequested) {
```
is always evaluated as `true` when testing, so in effect that change is a _no op_

I used different times and the results varies (hangs or none) but it's to really correlated to the times.

Reply:
Ok I get it : do you want me to remove this line or can we consider that in theory the cancellationToken could be already cancelled as you don't know from where it comes from ?

Reply:
I think the line can stay, it's logic not to add it if already cancelled.

However this makes both of your original changes _no ops_ which means nothing is solved. If you're getting different results it's because the test case is a bit too random in its results (not your fault but it makes things harder to see if the problem is fixed).

If this is related to the mono thread pool issue (like you cross-linked to) then it's unlikely that ay change in the handler code will be able to fix the issue.

Reply:
You are right, I have made more investigation and it seems that this is another behavior which causes the issue : https://github.com/xamarin/xamarin-macios/pull/5334#issuecomment-451132136